core/arch/arm/kernel/user_ta.c

```diff
@@ -824,6 +826,124 @@ static TEE_Result set_seg_prot(struct user_ta_ctx *utc,
 	return res;
 }
 
+#ifdef CFG_UNWIND
+
+/* An item in the exception index table */
+struct unwind_idx {
+	uint32_t offset;
+	uint32_t insn;
+};
+
+#define	EXIDX_CANTUNWIND 1
+
+static uint32_t offset_prel31(uint32_t addr, int32_t offset)
```

Comment:
What is this function doing?

Reply:
It computes `addr + offset` with the PREL31 encoding (relative address expressed as a signed 31-bit value). `addr` is a PREL31 value, manipulated as a `uint32_t`. The return value is a PREL31 too, returned as a `uint32_t` as well.

Reply:
The 31st bit can be 1 because `addr < 0` is a valid case. When copying a `.ARM.exidx` section that is before `.ARM.extab` and before the code section, to a place that is after, the positive relative addresses will become negative.
`addr` could be negative on input too, although I don't see this happening (would be if the linker puts `.ARM.exidx` after the code and/or `.ARM.extab`).

Maybe I misunderstood your question, if so, which change would you make here?

Reply:
31st bit is bit 30, it is part of the PREL31 encoding. I suppose you're talking about the 32nd bit (bit 31)?
Bit 31 is expected to be 0 indeed so if you prefer I can do this:

```C
static uint32_t offset_prel31(uint32_t addr, int32_t offset)
{
	return (addr + offset) & 0x7FFFFFFFUL;
}
```
...and return some error if `(item->offset & BIT32(31))` is set in `relocate_exidx()`.

Reply:
Yes, sorry for the confusion.